--- rt-n56u/trunk/user/busybox/busybox-1.24.x/scripts/kconfig/zconf.hash.c_shipped
+++ rt-n56u/trunk/user/busybox/busybox-1.24.x/scripts/kconfig/zconf.hash.c_shipped
@@ -156,7 +156,7 @@ static struct kconf_id_strings_t kconf_id_strings_contents =
 #ifdef __GNUC__
 __inline
 #endif
-struct kconf_id *
+static struct kconf_id *
 kconf_id_lookup (register const char *str, register unsigned int len)
 {
   enum
